### ISSUE TITLE
Org names display logic

### ISSUE THREAD

**Issue by mariongalley** [ThreeSixtyGiving/grantnav#821]
Please can you let me know what the logic is for the Org page names? Can you check whether it's the same logic as we use to display the Org name on the Grant Cards in the Search Results?

I don't think the current logic is suitable, for instance if you search Comic Relief in the Funders page, you are shown this org whose name is displayed as CHARITY PROJECTS, which doesn't look right. http://insightsng.vs.mythic-beasts.com:8888/org/GB-CHC-326568

**Comment by KDuerden** [comment]
For publishers/funders, the name should continue to be derived using the existing logic (funder name as provided in most recent grant) and not be based on external datasets.

**Comment by mariongalley** [comment]
@robredpath Yeah I appreciate it's technically a correct name for the org, but I think we'd like to change that logic to use a name that people will recognise before we push the overall org page update to live

**Comment by KDuerden** [comment]
From my perspective, a publishing funder does need an approach that respects what name they have put in their own data - given that all the grants that appear in GrantNav are there because of their having shared it.

For Charity Projects it is the most distinctive example of a registered name not matching the 'brand' name but there are lots of other funders with a variation of some kind.

There is also an issue of the government organisation Register which was retired by GDS but is still in use by government and lottery funders for their org IDs. The names on this register are not kept up to date and so pulling the name from this currently depreciating list provides incorrect information. However because there isn't a suitable replacement org ID list/register, we can't start moving those funders onto anything else.

https://findthatcharity.uk/orgid/GB-GOR-PB188

@stevieflow and I spent a long time supporting the National Lottery Community Fund (new name of BLF) to update their funding name to the new brand, and so appearing to roll back to an old name/brand in GrantNav would be a big issue for them.

**Comment by mariongalley** [comment]
@robredpath Thanks for that - I'm sorry I hadn't managed to convey sufficiently that although we display alternate names we'd still want the "best" name to be the main one displayed on the page.

I appreciate this will be a lot of work to change, but I do think it is worth it as publishers will be alienated by the big name at the top being "wrong" in their eyes. We can plan for this development to happen after the Analytics/Monitoring and Data Quality Tool work as discussed.

**Comment by robredpath** [comment]
Thanks both - that's really helpful, and I think it's clear now what needs to be done. 

As long as we can keep the fundamental model of there being record per organisation, and the same name being used on the org page and on org cards on the funders and recipients pages, the rework is much more tractable.

**Comment by michaelwood** [comment]
This is currently 'pre-requisite for live' but I'm not sure of its status or what the action needed here?

**Comment by KDuerden** [comment]
It is the naming logic currently used for name used in Funder pages and Funder filter values - which is the Funding Org:Name value in the most recent published grant.

**Comment by KDuerden** [comment]
This was the background - https://github.com/ThreeSixtyGiving/grantnav/issues/409

**Comment by kindly** [comment]
This is a bit more complicated then expressed above, as Org names have to be unified when the org is acting as a Funder and a Reciever.
Below is an attempt at the logic that "I think" is best based on the above logic when deciding on what the Org name should be.

1. If the organisation has ever acted as a **Funder** the Org name should be taken from "Funding Org:Name" from the latest grant, where they acted as a funder, accociated with *any* org-ids that are considered the same org.
2. If the organisation has only ever acted as a **Reciever** and at least one of the matching org-ids is found in FindThat charity then the name should be taken from  FindThatCharity.
3. If the oranisations org-ids can not be found in FindThatCharity then the name should be taken from the latest grant found for that org-id from the "Recipent Org:Name".

Points 2 and 3 have not been clarified from the above, and would like to check if there is agreement on them?  Most importantly should 2 be taken before 3? i.e should FindThatCharity names be taken before **Recipient** names that are supplied in the data?

If that is the logic needed, then there will be quite a bit of work to make that possible, as it is more complicated than the current approach which is farily complex already.

**Comment by mariongalley** [comment]
Have discussed this with Katherine, our thoughts are below.

> 1. If the organisation has ever acted as a Funder the Org name should be taken from "Funding Org:Name" from the latest grant, where they acted as a funder, accociated with any org-ids that are considered the same org.
> 2. If the organisation has only ever acted as a Reciever and at least one of the matching org-ids is found in FindThat charity then the name should be taken from FindThatCharity.
> 3. If the oranisations org-ids can not be found in FindThatCharity then the name should be taken from the latest grant found for that org-id from the "Recipent Org:Name".

Yes, that's the right logic.

> Points 2 and 3 have not been clarified from the above, and would like to check if there is agreement on them? Most importantly should 2 be taken before 3? i.e should FindThatCharity names be taken before **Recipient** names that are supplied in the data?

Yep, 2 should be taken before 3.

In terms of complexity, I'm surprised this is very complicated given most of this logic exists elsewhere in GrantNav. Can we not re-use some of that? There are canonical Org IDs and names in the datastore for both funders and recipients - see for instance #681 
- Point 1 is the logic that we already use for Funders in the current Funder pages.
- Point 2 is the logic that we already use for Recipients (with some limitations on the types of Org IDs we use - potentially fewer than what FindThatCharity offers?)
    - Do we need new logic for prioritising Org IDs when there are several, e.g. CHC > COH?
- Point 3 is already the logic we use already when Org ID is internal

**Comment by kindly** [comment]
> In terms of complexity, I'm surprised this is very complicated given most of this logic exists elsewhere in GrantNav. Can we not re-use some of that? There are canonical Org IDs and names in the datastore for both funders and recipients - see for instance #681

However, the datastore is not merging multiple organisations into a single one if they match in Find that charity.  In other words in the datastore there is a canonical name per Org ID, but not a canonical name across Org ID that match.  The complexity comes from the merging then reapplying the logic to all the grants that now appear to be the same organisation.

> 
>     * Point 1 is the logic that we already use for Funders in the current Funder pages.

Not exactly, if 2 or more funders are actually the same one due their org ids.

> 
>     * Point 2 is the logic that we already use for Recipients (with some limitations on the types of Org IDs we use - potentially fewer than what FindThatCharity offers?)

We have not got this logic after multiple orgs are merged.  That is more complicated as we need to get all the grants from all the merged orgs then apply this logic.

>       
>       * Do we need new logic for prioritizing Org IDs when there are several, e.g. CHC > COH?

There is no proritizing Org IDs all are treated the same. We are currently just showing a list of Org Ids if the URL shows any of them.

>     * Point 3 is already the logic we use already when Org ID is internal

Yes that logic is the same.  

However, the merging makes points 1 and 2 much more complicated, along with new logic around which name we should take from this hierarchy, as at the moment we treat Recipients and Funders separately when it comes to naming.

So, in summary, it is the merging that is complicated and means the current logic can not be easily reused.  Alongside having now to apply a hierarchy of naming depending on if they are a funder, recipient in find that charity or other recipient.

**Comment by kindly** [comment]
@mariongalley 

As far as understand, the datastore has no real concept of an Organisation as it looks at everything on a grant level.
For each grants recipient it also has a list of ORG ids that it considers the same and the name that publisher gives it along with the find-that-charity name (if it exists).

So the most useful thing to do in the datastore, I think, would be to add a field to each grant for both recipient and funder as to its cononical name defined in the above way.

At the moment the work is done on the export from the datastore, but that could be made simpler by adding those fields into the data.

However, to do this the same work will need to be done, as outlined above, with an added step of feeding back this connonical name to the additional grant data.  So I would say doing that would take 1-2 days more than the 3 days, but I would have to run that past @michaelwood as I am not all that familiar with the datastore.

**Comment by mariongalley** [comment]
Thanks @kindly. @michaelwood Could you confirm whether doing this in the datastore would indeed take an additional 1-2 days? We're really keen to ensure this is captured in the datastore so that we can get the benefits or deduplicating orgs at more than just the funder page level in future.

Also just to note that we're going to put this work on hold as we don't think it's a priority over other work that we have going on at the moment, but will aim to roll it into future projects that involve GrantNav, e.g. Grants to Individuals will likely require us to do some work on identifiers, or we have a Duplicates working group planned towards the end of the year.

**Comment by mariongalley** [comment]
See also https://github.com/ThreeSixtyGiving/insights-ng/issues/108

**Comment by mariongalley** [comment]
@michaelwood Katherine noticed that all Mercers Company published Funder names have ALL caps names in the title but all other funders don’t - why is this? Can it be corrected?

The funders affected are: The Charity of Sir Richard Whittington; Mercers' Charitable Foundation; Mercers' School Memorial Trust (Incorporating the Merrett Bequest); The Earl of Northampton's Charity; St Paul's Schools Foundation.

**Comment by michaelwood** [comment]
This happens when we work down the preferences list of  1) publisher name from 360G 2) FindThatCharity 3) Grant data, and use the preferred name from FindThatCharity which lists organisations as they are listed on the various registers (companies house, charity commission etc). e.g. [Mercers Company - charity comission](https://register-of-charities.charitycommission.gov.uk/charity-details/?regId=326340&subId=0)  source [Mercers Company - findthatcharity](https://findthatcharity.uk/orgid/GB-CHC-326340.json)

It can't be reliably corrected as we have no source for a non all-caps version.

**Comment by michaelwood** [comment]
> This happens when we work down the preferences list of 1) publisher name from 360G 2) FindThatCharity 3) Grant data, and use the preferred name from FindThatCharity which lists organisations as they are listed on the various registers (companies house, charity commission etc). e.g. [Mercers Company - charity comission](https://register-of-charities.charitycommission.gov.uk/charity-details/?regId=326340&subId=0) source [Mercers Company - findthatcharity](https://findthatcharity.uk/orgid/GB-CHC-326340.json)
> 
> It can't be reliably corrected as we have no source for a non all-caps version.

Noticed on [FTC web version](https://findthatcharity.uk/orgid/GB-CHC-326340) there is a non-all-caps version, maybe @drkane has a special way of doing this or a different data source that isn't currently exported.

**Comment by michaelwood** [comment]
Going to close this one now, I think the original issue is now fixed and we can raise any further issues in new reports.

**Comment by mariongalley** [comment]
Agreed @michaelwood, I've made a note to ask David about an FTC based fix - will open a new issue if you need to do anything

**Comment by drkane** [comment]
@michaelwood Here's the code used for the titlecasing in find that charity: https://github.com/drkane/find-that-charity/blob/main/findthatcharity/utils.py#L106-L124

It uses the [titlecase](https://pypi.org/project/titlecase/) library along with a [custom function that deals with exceptions](https://github.com/drkane/find-that-charity/blob/main/findthatcharity/utils.py#L17-L103).

It's not perfect but I think it does a decent job on UK Charity names. There's [some tests showing the kind of thing it should work on](https://github.com/drkane/find-that-charity/blob/main/findthatcharity/tests.py#L19-L50).

I think the two options are probably:

 - I add the titlecase version to the FTC CSV output and we use that instead
 - the titlecase function is included in grantnav/the datastore and run on the data

**Comment by mariongalley** [comment]
@drkane re: two options, what do you think is preferable? Intuitively it seems to me more robust to address this at FTC export level rather than Datastore level

**Comment by drkane** [comment]
@mariongalley I think you're right that it would definitely be easier and wouldn't require any changes to the datastore, and only a minor one to Find that Charity. My only worry is that the titlecase algorithm doesn't get every name right, which is why I haven't used it as the default in the export from find that charity. But hopefully this'll be less of a problem anyway - the Charity Commission at least seems to be moving towards allowing lowercase letters.

**Comment by drkane** [comment]
@mariongalley I think you're right that it would definitely be easier and wouldn't require any changes to the datastore, and only a minor one to Find that Charity. My only worry is that the titlecase algorithm doesn't get every name right, which is why I haven't used it as the default in the export from find that charity. But hopefully this'll be less of a problem anyway - the Charity Commission at least seems to be moving towards allowing lowercase letters.
